arm9/source/utils/scripting.c

```diff
@@ -1624,57 +1669,89 @@ bool ValidateText(const char* text, u32 len) {
+            else cmt_start_chars = chars_in_bytes(ptr + off_disp_bytes, cmt_start_bytes);
         }
-        if (cmt_start <= 0) color_text = script_color_comment;
 
         // build text string
-        snprintf(txtstr, sizeof(txtstr), "%-*.*s", (int) TV_LLEN_DISP, (int) TV_LLEN_DISP, "");
-        if (ncpy) memcpy(txtstr, ptr + off_disp, ncpy);
+        snprintf(txtstr, sizeof(txtstr), "%-*.*s", (int) (TV_LLEN_DISP * MAX_CHAR_SIZE), (int) (TV_LLEN_DISP * MAX_CHAR_SIZE), "");
+        if (ncpy_bytes) {
+            memcpy(txtstr, ptr + off_disp_bytes, ncpy_bytes);
+        }
         for (char* d = txtstr; *d; d++) if (*d < ' ') *d = ' ';
-        if (al) memcpy(txtstr + p_al, al_str, strlen(al_str));
-        if (ar) memcpy(txtstr + p_ar, ar_str, strlen(ar_str));
+        if (ar) {
+            char* textstr_end = txtstr + ncpy_bytes;
+            u32 txtstr_ar_bytes = 0;
+            for (int txtstr_ar_i = 0; txtstr_ar_i < (int) strlen(ar_str); txtstr_ar_bytes += GetPrevCharSize(textstr_end - txtstr_ar_bytes), ++txtstr_ar_i);
+            memcpy(textstr_end - txtstr_ar_bytes, ar_str, sizeof(ar_str));
+        }
+        if (al) {
+            u32 txtstr_al_bytes = bytes_in_chars_u32(txtstr, strlen(al_str));
+            memmove(txtstr + strlen(al_str), txtstr + txtstr_al_bytes, sizeof(txtstr) - txtstr_al_bytes);
+            memcpy(txtstr, al_str, strlen(al_str));
+        }
 
         // draw line number & text
         DrawString(TOP_SCREEN, txtstr, x_txt, y, color_text, COLOR_STD_BG);
         if (TV_LNOS > 0) { // line number
-            if (ptr != ptr_next)
-                DrawStringF(TOP_SCREEN, x_lno, y, ((ptr == text) || (*(ptr-1) == '\n')) ? COLOR_TVOFFS : COLOR_TVOFFSL, COLOR_STD_BG, "%0*lu", TV_LNOS, nln);
+            bool prev_ww_line_full = ww && ww == chars_between_pointers(line_seek_chars(text, len, ww, ptr, -1), ptr);
+            bool last_line_empty = (ptr == text + len && (!len || ptr[-1] == '\n' || prev_ww_line_full) && !last_empty_line_drawn);
+            if (ptr != ptr_next || last_line_empty) {
+                DrawStringF(TOP_SCREEN, x_lno, y, ((ptr == text) || (ptr[-1] == '\n')) ? COLOR_TVOFFS : COLOR_TVOFFSL, COLOR_STD_BG, "%0*lu", TV_LNOS, nln);
+                if (last_line_empty) last_empty_line_drawn = true;
+            }
             else DrawStringF(TOP_SCREEN, x_lno, y, COLOR_TVOFFSL, COLOR_STD_BG, "%*.*s", TV_LNOS, TV_LNOS, " ");
         }
 
+        if (cursor) {
+            u32 cursor_line_offset_chars = chars_between_pointers(ptr + off_disp_bytes, cursor);
+            if (cursor >= ptr + off_disp_bytes && cursor <= ptr + off_disp_bytes + ncpy_bytes && cursor_line_offset_chars < TV_LLEN_DISP
+                    && (cursor != ptr + off_disp_bytes + ncpy_bytes || is_newline(cursor) || cursor == text + len)) {
+                DrawRectangle(TOP_SCREEN, x_txt + cursor_line_offset_chars * FONT_WIDTH_EXT, y, FONT_WIDTH_EXT, 1, COLOR_RED);
```

Comment:
Without a blinking line, a hollow red rect is the next best cursor I could think of. These coords are picked so that they are drawn over by text.